### PR TITLE
doc: nastpec and documentation

### PR DESCRIPTION
**Pull request by deluca-mike**

### CODE REVIEW

File: src/ContractHelper.sol [hunk 1]
```diff
@@ -2,26 +2,47 @@
 
 pragma solidity 0.8.21;
 
+/**
+ * @title A helper contract to get the address of a contract deployed by aan account with a given nonce.
```
Comment (PierrickGT) [inline src/ContractHelper.sol:6]:
```suggestion
 * @title A helper contract to get the address of a contract deployed by an account with a given nonce.
```

File: src/SignatureChecker.sol [hunk 2]
```diff
@@ -4,6 +4,7 @@ pragma solidity 0.8.21;
 
 import { IERC1271 } from "./interfaces/IERC1271.sol";
 
+/// @title A library to handle ECDSA/secp256k1 and ERC1271 signatures, individually or in arbitrarily in combination.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:7]:
```suggestion
/// @title A library to handle ECDSA/secp256k1 and ERC1271 signatures, individually or arbitrarily in combination.
```

**Review by toninorair — APPROVED** [review]

File: src/SignatureChecker.sol [hunk 3]
```diff
@@ -65,6 +103,13 @@ library SignatureChecker {
             abi.decode(result, (bytes32)) == bytes32(IERC1271.isValidSignature.selector);
     }
 
+    /**
+     * @dev    Returns the signer of an ECDSA/secp256k1 signature for some digest.
+     * @param  digest    The hash of the data that was signed.
+     * @param  signature A byte array ECDSA/secp256k1 signature.
+     * @return error     An error, if any, that occurred during the singer recovery.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:110]:
```suggestion
     * @return error     An error, if any, that occurred during the signer recovery.
```

File: src/SignatureChecker.sol [hunk 4]
```diff
@@ -76,6 +121,15 @@ library SignatureChecker {
         return recoverECDSASigner(digest, v, r, s);
     }
 
+    /**
+     * @dev    Returns the signer of an ECDSA/secp256k1 signature for some digest.
+     * @param  digest The hash of the data that was signed.
+     * @param  v      An ECDSA/secp256k1 signature parameter.
+     * @param  r      An ECDSA/secp256k1 signature parameter.
+     * @param  s      An ECDSA/secp256k1 signature parameter.
+     * @return error  An error, if any, that occurred during the singer recovery.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:130]:
```suggestion
     * @return error  An error, if any, that occurred during the signer recovery.
```

File: src/SignatureChecker.sol [hunk 5]
```diff
@@ -94,6 +148,13 @@ library SignatureChecker {
         return (signer == address(0)) ? (Error.InvalidSignature, address(0)) : (Error.NoError, signer);
     }
 
+    /**
+     * @dev    Returns an error, if any, in validating an ECDSA/secp256k1 signature for a signer and digest.
+     * @param  signer    The address of the account purported to have signed.
+     * @param  digest    The hash of the data that was signed.
+     * @param  signature A byte array ERC1271 signature.
+     * @return error     An error, if any, that occurred during the singer recovery.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:156]:
```suggestion
     * @return error     An error, if any, that occurred during the signer recovery.
```

File: src/SignatureChecker.sol [hunk 6]
```diff
@@ -104,6 +165,15 @@ library SignatureChecker {
...
     }
 
+    /**
+     * @dev    Returns an error, if any, in validating an ECDSA/secp256k1 signature for a signer and digest.
+     * @param  signer The address of the account purported to have signed.
+     * @param  digest The hash of the data that was signed.
+     * @param  v      An ECDSA/secp256k1 signature parameter.
+     * @param  r      An ECDSA/secp256k1 signature parameter.
+     * @param  s      An ECDSA/secp256k1 signature parameter.
+     * @return error  An error, if any, that occurred during the singer recovery.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:175]:
```suggestion
     * @return error  An error, if any, that occurred during the signer recovery.
```

File: src/SignatureChecker.sol [hunk 7]
```diff
@@ -116,6 +186,12 @@ library SignatureChecker {
         return (recoverError == Error.NoError) ? validateRecoveredSigner(signer, recoveredSigner) : recoverError;
     }
 
+    /**
+     * @dev    Returns an error if `signer` is not `recoveredSigner`.
+     * @param  signer          The address of the some signer.
+     * @param  recoveredSigner The address of the some recoveredSigner.
```
Comment (PierrickGT) [inline src/SignatureChecker.sol:192]:
```suggestion
     * @param  signer          The address of some signer.
     * @param  recoveredSigner The address of some recoveredSigner.
```

File: src/StatefulERC712.sol [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+/// @dev   An abstract implementation to satisfy stateful EIP-712 with nonces.
+abstract contract StatefulERC712 is IStatefulERC712, ERC712 {
+    mapping(address account => uint256 nonce) internal _nonces; // Nonces for all signatures.
+
+    constructor(string memory name_) ERC712(name_) {}
+
+    /******************************************************************************************************************\
+    |                                       External/Public View/Pure Functions                                        |
+    \******************************************************************************************************************/
+
```
Comment (PierrickGT) [inline src/StatefulERC712.sol:19]: Add
```suggestion
/// @inheritdoc IStatefulERC712
```

File: src/interfaces/IERC20.sol [hunk 9]
```diff
@@ -2,38 +2,102 @@
...
+     * @notice Allows a calling account to approve `spender` to spend up to `amount` of its token balance.
+     * @param  spender The address of the account being allowed to spend up to the allowed amount.
+     * @param  amount  The amount of the allowance being approved.
+     * @return success Whether or not the approval was successful.
+     */
     function approve(address spender, uint256 amount) external returns (bool success);
 
+    /**
+     * @notice Allows a calling account to decrease the allowance `spender` has, by `subtractedAmount`.
+     * @param  spender          The address of the account being who allowance is being decreased.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:42]:
```suggestion
     * @param  spender          The address of the account whose allowance is being decreased.
```

File: src/interfaces/IERC20.sol [hunk 10]
```diff
@@ -2,38 +2,102 @@
...
+     * @param  subtractedAmount The amount the allowance is being decreased by.
+     * @return success          Whether or not the decrease in allowance was successful.
+     */
     function decreaseAllowance(address spender, uint256 subtractedAmount) external returns (bool success);
 
+    /**
+     * @notice Allows a calling account to increase the allowance `spender` has, by `addedAmount`.
+     * @param  spender     The address of the account being who allowance is being increased.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:50]:
```suggestion
     * @param  spender     The address of the account whose allowance is being increased.
```

File: src/interfaces/IERC20.sol [hunk 11]
```diff
@@ -2,38 +2,102 @@
...
+     * @param  addedAmount The amount the allowance is being increased by.
+     * @return success     Whether or not the increase in allowance was successful.
+     */
     function increaseAllowance(address spender, uint256 addedAmount) external returns (bool success);
 
+    /**
+     * @notice Allows a calling account to transfer `amount` tokens to `recipient`.
+     * @param  recipient The address of the recipient who's token balance will be incremented.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:58]:
```suggestion
     * @param  recipient The address of the recipient whose token balance will be incremented.
```

File: src/interfaces/IERC20.sol [hunk 12]
```diff
@@ -2,38 +2,102 @@
...
+     * @param  amount    The amount of tokens being transferred.
+     * @return success   Whether or not the transfer was successful.
+     */
     function transfer(address recipient, uint256 amount) external returns (bool success);
 
+    /**
+     * @notice Allows a calling account to transfer `amount` tokens from `sender`, with allowance, to a `recipient`.
+     * @param  sender    The address of the sender who's token balance will be decremented.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:66]:
```suggestion
     * @param  sender    The address of the sender whose token balance will be decremented.
```

File: src/interfaces/IERC20.sol [hunk 13]
```diff
@@ -2,38 +2,102 @@
...
+     * @param  recipient The address of the recipient who's token balance will be incremented.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:67]:
```suggestion
     * @param  recipient The address of the recipient whose token balance will be incremented.
```

File: src/interfaces/IERC20.sol [hunk 14]
```diff
@@ -2,38 +2,102 @@
...
+     */
     function transferFrom(address sender, address recipient, uint256 amount) external returns (bool success);
 
     /******************************************************************************************************************\
     |                                              View/Pure Functions                                                 |
     \******************************************************************************************************************/
 
+    /**
+     * @notice Returns the allowance `spender` is allowed to spend on behalf of `account`.
+     * @param  account   The address of the account who's token balance `spender` is allowed to spend.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:79]:
```suggestion
     * @param  account   The address of the account whose token balance `spender` is allowed to spend.
```

File: src/interfaces/IERC20.sol [hunk 15]
```diff
@@ -2,38 +2,102 @@
...
     function allowance(address account, address spender) external view returns (uint256 allowance);
 
+    /**
+     * @notice Returns the token balance of `account`.
+     * @param  account The address of some account.
+     * @return balance The token balance of `account`.
+     */
     function balanceOf(address account) external view returns (uint256 balance);
 
+    /// @notice Returns the number of decimals UIs should assume all amounts have.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:92]: Add return param doc.

File: src/interfaces/IERC20.sol [hunk 16]
```diff
@@ -2,38 +2,102 @@
...
     function decimals() external view returns (uint8 decimals);
 
+    /// @notice Returns the name of the contract/token.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:95]: Add return param doc.

File: src/interfaces/IERC20.sol [hunk 17]
```diff
@@ -2,38 +2,102 @@
...
     function name() external view returns (string memory name);
 
+    /// @notice Returns the symbol of the token.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:98]: Add return param doc.

File: src/interfaces/IERC20.sol [hunk 18]
```diff
@@ -2,38 +2,102 @@
...
     function symbol() external view returns (string memory symbol);
 
+    /// @notice Returns the current total supply of the token.
```
Comment (PierrickGT) [inline src/interfaces/IERC20.sol:101]: Add return param doc.

File: src/interfaces/IERC20Permit.sol [hunk 19]
```diff
@@ -3,26 +3,41 @@
...
-import { IERC712 } from "./IERC712.sol";
+import { IStatefulERC712 } from "./IStatefulERC712.sol";
 
-interface IERC20Permit is IERC20, IERC712 {
+/// @title Permit Extension for ERC20 Signed Approvals via EIP-712 with EIP-2612 and EIP-1271 compatibility.
+/// @dev   The interface as defined by EIP-2612: https://eips.ethereum.org/EIPS/eip-2612
+interface IERC20Permit is IERC20, IStatefulERC712 {
+    /**
+     * @notice Approves `spender` to spend up to `amount` of the token balance of `owner`, via a signature.
+     * @param  owner    The address of the account who's token balance is being approved to be spent by `spender`.
```
Comment (PierrickGT) [inline src/interfaces/IERC20Permit.sol:13]:
```suggestion
     * @param  owner    The address of the account whose token balance is being approved to be spent by `spender`.
```

File: src/interfaces/IERC20Permit.sol [hunk 20]
```diff
@@ -3,26 +3,41 @@
...
+     * @param  spender  The address of an account allowed to spend on behalf of `owner`.
+     * @param  value    The amount of the allowance being approved.
+     * @param  deadline The last block number where the signature is still valid.
```
Comment (PierrickGT) [inline src/interfaces/IERC20Permit.sol:16]: `deadline` is the `block.timestamp` and not the `block.number` no?

File: src/interfaces/IERC20Permit.sol [hunk 21]
```diff
@@ -3,26 +3,41 @@
...
-    function permit(
-        address account,
-        address spender,
-        uint256 amount,
-        uint256 deadline,
-        bytes memory signature
-    ) external;
+    /**
+     * @notice Approves `spender` to spend up to `amount` of the token balance of `owner`, via a signature.
+     * @param  owner     The address of the account who's token balance is being approved to be spent by `spender`.
```
Comment (PierrickGT) [inline src/interfaces/IERC20Permit.sol:33]:
```suggestion
     * @param  owner     The address of the account whose token balance is being approved to be spent by `spender`.
```

File: src/interfaces/IERC20Permit.sol [hunk 22]
```diff
@@ -3,26 +3,41 @@
...
+     * @param  spender   The address of an account allowed to spend on behalf of `owner`.
+     * @param  value     The amount of the allowance being approved.
+     * @param  deadline  The last block number where the signature is still valid.
+     * @param  signature An arbitrary signature (EIP-712).
+     */
+    function permit(address owner, address spender, uint256 value, uint256 deadline, bytes memory signature) external;
 
+    /// @notice Returns the EIP712 typehash used in the encoding of the digest for the permit function.
```
Comment (PierrickGT) [inline src/interfaces/IERC20Permit.sol:41]: Add return param doc.

File: src/interfaces/IStatefulERC712.sol [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+pragma solidity 0.8.21;
+
+import { IERC712 } from "./IERC712.sol";
+
+/// @title Stateful Extension for EIP-712 typed structured data hashing and signing with nonces.
+interface IStatefulERC712 is IERC712 {
+    /**
+     * @notice Revert message when a signing account's nonce is reused by a signature.
+     * @param  nonce        The nonce used in the signature.
+     * @param  currentNonce The last nonce used in a signature by then signing account.
```
Comment (PierrickGT) [inline src/interfaces/IStatefulERC712.sol:12]:
```suggestion
     * @param  currentNonce The last nonce used in a signature by the signing account.
```

**Review by PierrickGT — CHANGES_REQUESTED** [review]

**Pull request merged**